File: LumiMind-Backend/src/models/GameResult.js
//
// GameResult.js
//
// Mongoose schema matching LumiMind/Models/GameResult.swift field-for-field.
// See API_CONTRACT.md for the canonical JSON shape returned to clients.
//
// Converted to ES Modules in Build Prompt #3 for consistency with the
// rest of the backend (was CommonJS in Build Prompt #2).
//

import mongoose from 'mongoose';

// Locked list — mirrors GameName.swift. Do not add/rename without
// updating GameResult.swift and API_CONTRACT.md.
export const GAME_NAMES = [
  'Memory Matrix',
  'Speed Match',
  'Lost in Migration',
  'Brain Shift',
  'Pirate Passage',
  'Splitting Seeds',
];

// Locked list — mirrors GameCategory.rawValue in DesignSystem.swift.
// Do not add/rename without updating DesignSystem.swift and API_CONTRACT.md.
export const CATEGORIES = [
  'Speed',
  'Memory',
  'Attention',
  'Flexibility',
  'Problem Solving',
  'Math',
];

// Locked mapping from display category name -> categoryScores camelCase
// key. Mirrors GameCategory.categoryScoresKey in GameResult.swift exactly
// (see API_CONTRACT.md's canonical category table). This is the ONLY
// place this mapping is defined on the backend — do not re-derive it
// in controllers or services.
export const CATEGORY_TO_SCORE_KEY = {
  Speed: 'speed',
  Memory: 'memory',
  Attention: 'attention',
  Flexibility: 'flexibility',
  'Problem Solving': 'problemSolving',
  Math: 'math',
};

const gameResultSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  gameName: {
    type: String,
    required: true,
    enum: GAME_NAMES,
  },
  category: {
    type: String,
    required: true,
    enum: CATEGORIES,
  },
  score: {
    type: Number,
    required: true,
  },
  durationSeconds: {
    type: Number,
    required: true,
  },
  isFitTest: {
    type: Boolean,
    default: false,
  },
  playedAt: {
    type: Date,
    default: Date.now,
  },
});

const GameResult = mongoose.model('GameResult', gameResultSchema);

export default GameResult;

File: LumiMind/Models/GameResult.swift
//
//  GameResult.swift
//  LumiMind
//
//  Codable model matching the backend `GameResult` document exactly.
//  See LumiMind-Backend/API_CONTRACT.md for the canonical JSON shape.
//

import Foundation

// MARK: - GameResult

struct GameResult: Codable, Identifiable, Equatable {
    let id: String
    let userId: String
    let gameName: String
    let category: String
    let score: Int
    let durationSeconds: Int
    let isFitTest: Bool
    let playedAt: Date

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case userId
        case gameName
        case category
        case score
        case durationSeconds
        case isFitTest
        case playedAt
    }
}

// MARK: - GameName
//
// The six locked games. Each maps to exactly one `GameCategory`
// (defined in DesignSystem.swift) so `category` is always derived
// from `gameName`, never typed separately by call sites.

enum GameName: String, CaseIterable, Codable {
    case memoryMatrix = "Memory Matrix"
    case speedMatch = "Speed Match"
    case lostInMigration = "Lost in Migration"
    case brainShift = "Brain Shift"
    case piratePassage = "Pirate Passage"
    case splittingSeeds = "Splitting Seeds"

    /// Locked game -> category mapping. Splitting Seeds -> Math, etc.
    var category: GameCategory {
        switch self {
        case .memoryMatrix:      return .memory
        case .speedMatch:        return .speed
        case .lostInMigration:   return .attention
        case .brainShift:        return .flexibility
        case .piratePassage:     return .problemSolving
        case .splittingSeeds:    return .math
        }
    }
}

// MARK: - GameCategory + API naming

extension GameCategory {
    /// The exact string sent/received as `GameResult.category` and used
    /// in the `category` gradient names. This IS `rawValue` — exposed
    /// here under an explicit name so call sites are self-documenting
    /// about which contract they're satisfying.
    var apiCategoryName: String { rawValue }

    /// camelCase key matching `User.categoryScores` JSON keys exactly
    /// (e.g. "Problem Solving" -> "problemSolving"). This is the ONLY
    /// place this mapping is defined — do not re-derive it elsewhere.
    var categoryScoresKey: String {
        switch self {
        case .speed:          return "speed"
        case .memory:         return "memory"
        case .attention:      return "attention"
        case .flexibility:    return "flexibility"
        case .problemSolving: return "problemSolving"
        case .math:            return "math"
        }
    }
}